add docs for Rails instrumentation (#103)
signalfx/splunk-otel-ruby: docs/instrumenting-rails.md

# Instrumenting Rails

> TODO

Reviewer: @tsloughter-splunk Any Rails application can be instrumented? Any conflict if other libraries are instrumented? Just asking because the behavior for Django apps in Python required going through lots of hoops and exceptions.

Reviewer 2: I wouldn't expect anything like that. Do you have any examples for the common difficulties in Django?

Reviewer: Lots of small caveats, but perhaps not applicable here: https://docs.splunk.com/Observability/gdi/get-data-in/application/python/instrumentation/instrument-python-frameworks.html

Reviewer 2: I see some similarities to ruby setups from 10 years ago, not so much the modern ones, so I'd say let's just release this now, and if customers come with questions, we'll update the docs based on questions.

``` ruby
gem "opentelemetry-instrumentation-rails", "~> 0.24"
```

> or

Reviewer: ```suggestion
or install the gem using `bundle`:
```

```shell
bundle add opentelemetry-instrumentation-rails --version "~> 0.24"
```

> Configure OpenTelemetry to use all available instrumentation libraries using the

Reviewer: ```suggestion
Configure OpenTelemetry to use all available instrumentation libraries by adding the
```

Reviewer 2: For me one thing that's missing is where to put this code, and even OTel's readme doesn't explain that. One logical place would be `config/initializers`, but I'm familiar with all intricacies of hooking and I don't know if that will work for all cases - though I usually test it this way, and I've never had any issues.

Reviewer 2: Closest possible place to what they have in the [example](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/main/instrumentation/rails/example/trace_request_demonstration.ru), ie. before `Rails.application.initialize!`, would be `config/environment`, although it feels awkward.

Author: The Otel Ruby docs mention doing it in initializers, https://opentelemetry.io/docs/instrumentation/ruby/automatic/ -- I've updated this doc to also mention you'd put this in an initializer for Rails.

`Splunk::Otel` module from `splunk/otel` and `use_all()` method:

``` ruby
require "splunk/otel"
...
Splunk::Otel.configure do |c|
  c.use_all()
end
```

You can disable individual components' instrumentation as options to
`use_all`. For example to disable Active Record instrumentation:

``` ruby
Splunk::Otel.configure do |c|
  c.use_all({ 'OpenTelemetry::Instrumentation::ActiveRecord' => { enabled: false } })
end
```

> To enable only Rails an indiviual `c.use` can be used:

Reviewer: ```suggestion
To enable only Rails, you can use an indiviual `c.use`:
```

# Instrumenting Rails

Install the instrumentation library by adding it to your project's `Gemfile`:

``` ruby
gem "opentelemetry-instrumentation-rails", "~> 0.24"
```

or install the gem using `bundle`:

```shell
bundle add opentelemetry-instrumentation-rails --version "~> 0.24"
```

Configure OpenTelemetry to use all available instrumentation libraries by adding the
`Splunk::Otel` module from `splunk/otel` and `use_all()` method to a Rails initializer:

``` ruby

> # config/initializers/opentelemetry.rb

Reviewer 2: I think it's safer to go with `config/environment.rb` above `Rails.application.initialize!`.

Author: Why's that? I think we should have a really good reason to differ from the upstream docs on something like this. Happy to change it, just want to make sure.

Author: @jtmal-signalfx ping

Reviewer 2: oops! somehow I missed the notification. Alright, we can follow upstream.
